src/plugins/lua/replies.lua

```diff
@@ -91,16 +98,146 @@ local function replies_check(task)
+    return nil
+  end
+
+  local function configure_redis_scripts()
+    local redis_script_zadd_global = [[
+      local global_size = redis.call('ZCARD', KEYS[1])
+      if global_size > {= mgs =} then
+        redis.call('ZREMRANGEBYRANK', KEYS[1], '{= mgs =}', global_size)
+      end
+      -- adding recipients to the global replies set
+      redis.call('ZADD', KEYS[1], ARGV[1], ARGV[2])
+      ]]
+    local set_script_zadd_global = lua_util.jinja_template(redis_script_zadd_global,
+              { mgs = settings['max_global_size'] })
+    global_replies_set_script =  lua_redis.add_redis_script(set_script_zadd_global, redis_params)
+
+    local redis_script_zadd_local = [[
+      local local_replies_set_size = redis.call('ZCARD', KEYS[1])
+      if local_replies_set_size > {= mls =} then
+        redis.call('ZREMRANGEBYRANK', KEYS[1], '{= mls =}', local_replies_set_size)
+      end
+      -- adding recipients for the local replies set
+      redis.call('ZADD', KEYS[1], ARGV[1], ARGV[2])
+      -- setting expire for local replies set
+      redis.call('EXPIRE', KEYS[1], tostring(math.floor('{= exp =}')))
+    ]]
+    local set_script_zadd_local = lua_util.jinja_template(redis_script_zadd_local,
+      { exp = settings['expire'], mls = settings['max_local_size'] })
+    local_replies_set_script = lua_redis.add_redis_script(set_script_zadd_local, redis_params)
+  end
+
+  local function add_to_global_replies_set(recipients, global_key, task_time)
+    lua_util.debugm(N, task, 'Adding recipients %s to global replies set', recipients)
+
+    local function zadd_global_set_cb(err, data)
+      if err ~= nil then
+        rspamd_logger.errx(task, 'failed to add recipients %s to global replies set with error: %s', recipients, err)
+        return
+      end
+      rspamd_logger.infox(task, 'added recipients %s to global set', recipients)
+    end
+    for _, rcpt in ipairs(recipients) do
+      lua_redis.exec_redis_script(global_replies_set_script,
+              { task=task, is_write = true },
+              zadd_global_set_cb,
+              { global_key },
+              { task_time, rcpt })
+    end
+  end
+
+  local function update_global_replies_set(recipients, sender_key, global_key, task_time)
+    local function redis_zrange_cb(err, data)
+      if err ~= nil then
+        rspamd_logger.errx(task,
+                'redis_zrange_cb error when reading zrange withscores from global replies set with error: %s', err)
+        return
+      end
+      local last_score = tonumber(data[#data])
+      lua_util.debugm(N, task, 'last score %s of global replies set was received', last_score)
+
+      -- if last score wasn't found
+      if last_score == nil then
+        lua_util.debugm(N, task,
+                'have not found any senders in global replies set, considering last score as 0')
+        last_score = 0
+      end
+
+      -- updating params considering last score of existing sender
+      task_time = task_time + last_score
```

Comment:
Is there a reason we don't rather increment it by 1 each time?

Could we please get a more detailed description for this PR if possible to help me understand the intentions?

Reply:
Probably it is so newer data would be retained rather than older data, that's reasonable ...